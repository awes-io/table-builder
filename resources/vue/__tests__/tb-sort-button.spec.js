import { shallowMount } from '@vue/test-utils'
import tbSortButton from './../tb-sort-button'

describe('tb-sort-button', () => {

    let wrapper;

    let state = null;

    beforeEach(() => {

        state = null

        wrapper = shallowMount(tbSortButton, {
            propsData: {
                asc: 'asc',
                desc: 'desc'
            },
            slots: {
                default: '<div class="slot-content"></div>'
            },
            mocks: {
                $route: {
                    path: 'path',
                    query: {}
                },
                $get: () => {},
                $router: {
                    $setParam: (data) => state = data
                }
            }
        });
    });

    it('renders slot inside of the button', () => {

        const button = wrapper.find('button')

        expect(button.findAll('div.slot-content').length).toEqual(1)
    })

    it('renders only static class when order doesn\'t match', () => {

        expect(wrapper.find('button').classes()).toEqual(['tb-sort-button'])
    })

    it('renders ascending and active classes when order matches', () => {

        wrapper.vm.$route.query = { orderBy: 'asc' }

        expect(wrapper.find('button').classes()).toContain('is-active')

        expect(wrapper.find('button').classes()).toContain('is-asc')
    })

    it('renders descending and active classes when order matches', () => {

        wrapper.vm.$route.query = { orderBy: 'desc' }

        expect(wrapper.find('button').classes()).toContain('is-active')

        expect(wrapper.find('button').classes()).toContain('is-desc')
    })

    it('iterates through asc, desc and empty value on button click', () => {

        const button = wrapper.find('button')

        expect(state).toBeNull()

        button.trigger('click')

        expect(state).toEqual({ orderBy: 'asc' })

        wrapper.vm.$route.query = { orderBy: 'asc' } // mock router behaviour
        button.trigger('click')

        expect(state).toEqual({ orderBy: 'desc' })

        wrapper.vm.$route.query = { orderBy: 'desc' } // mock router behaviour
        button.trigger('click')

        expect(state).toEqual({ orderBy: null })
    })

     it('iterates through one passed order prop and empty value on button click, when other order prop is empty', () => {

        wrapper.setProps({ desc: '' })

        const button = wrapper.find('button')

        expect(state).toBeNull()
        expect(wrapper.find('button').classes()).toContain('is-desc')

        button.trigger('click')

        expect(state).toEqual({ orderBy: 'asc' })

        wrapper.vm.$route.query = { orderBy: 'asc' } // mock router behaviour
        button.trigger('click')

        expect(state).toEqual({ orderBy: '' })
    })

    it('matches overall snapshot', () => {

        expect(wrapper.html()).toMatchSnapshot();
    })
})