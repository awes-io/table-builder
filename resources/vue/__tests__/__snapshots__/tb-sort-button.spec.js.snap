// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tb-sort-button matches overall snapshot 1`] = `"<button class=\\"tb-sort-button\\"><div class=\\"slot-content\\"></div> <svg width=\\"16\\" height=\\"16\\" viewBox=\\"0 0 20 20\\"><path d=\\"M17 13H3l7-8z\\" style=\\"display: none;\\"></path> <path d=\\"M3 7h14l-7 8z\\" style=\\"display: none;\\"></path> <path d=\\"M17 9H3l7-8z M3 11h14l-7 8z\\"></path></svg></button>"`;
